File: solutions/haskell/bob/1/src/Bob.hs
module Bob (responseFor) where

import qualified Data.Text as T
import           Data.Text (Text)
import           Data.Char (isAlpha)

responseFor :: Text -> Text
responseFor xs
  | isSilent xs = T.pack "Fine. Be that way!"
  | isYelling xs && isAsking xs = T.pack "Calm down, I know what I'm doing!"
  | isYelling xs = T.pack "Whoa, chill out!"
  | isAsking xs = T.pack "Sure."
  | otherwise = T.pack "Whatever."

isSilent :: Text -> Bool
isSilent xs = T.stripStart xs == T.empty

isYelling :: Text -> Bool
isYelling xs = T.any isAlpha xs && xs == T.toUpper xs 

isAsking :: Text -> Bool
isAsking xs = T.isSuffixOf (T.pack "?") (T.stripEnd xs)

File: solutions/cpp/bob/1/bob.cpp
#include "bob.h"

#include <algorithm>
#include <locale>

namespace bob {
    std::string hey(std::string input) {
        const auto is_question{
            input[input.find_last_not_of("\n\r\t ")] == '?'
        };
        const auto is_yelling{
            std::all_of(
                input.begin(),
                input.end(),
                [](char c){ return std::isupper(c) || !std::isalpha(c); }
            ) && std::any_of(
                input.begin(),
                input.end(),
                [](char c){ return std::isupper(c); }
            )
        };
        const auto is_silent{
            std::all_of(
                input.begin(),
                input.end(),
                [](char c){ return std::isspace(c); }
            )
        };
        if (!is_yelling && is_question) {
            return "Sure.";
        }
        if (is_yelling && !is_question) {
            return "Whoa, chill out!";
        }
        if (is_yelling && is_question) {
            return "Calm down, I know what I'm doing!";
        }
        if (is_silent) {
            return "Fine. Be that way!";
        }
        return "Whatever.";
    }
}
